layers/+lang/ruby/funcs.el

```diff
@@ -22,11 +22,55 @@
 
 
 
-;; backend
+;; macros
+
+(defmacro spacemacs|eval-for-enabled-ruby-mode (body)
+  "Evaluate BODY for the currently active ruby mode.
+
+All occurrences of `mode' in BODY are replaced with the actual mode name symbol.
+All occurrences of `hook' in BODY are replaced with the actual hook name symbol.
+All occurrences of `local-vars-hook' in BODY are replaced with the actual local
+vars hook name symbol.
+Note that `mode', `hook' and `local-vars-hook' need to be explicitly quoted
+except when they are themselves inside a macro."
+  (declare (debug (form)) (indent 0))
+  (if ruby-enable-enh-ruby-mode
+      ;; enh-ruby-mode
+      (progn
+        (let ((new 'enh-ruby-mode)
+              (old 'mode))
+          (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))
+        (let ((new 'enh-ruby-mode-hook)
+              (old 'hook))
+          (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))
+        (let ((new 'enh-ruby-mode-local-vars-hook)
+              (old 'local-vars-hook))
+          (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body)))
+    ;; ruby-mode
+    (let ((new 'ruby-mode)
+          (old 'mode))
+      (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))
+    (let ((new 'ruby-mode-hook)
+          (old 'hook))
+      (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))
+    (let ((new 'ruby-mode-local-vars-hook)
+          (old 'local-vars-hook))
+      (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body))))
+
+(defun spacemacs//ruby-substitute-predicate (body)
+  "Predicate to replace all `old' occurrences by `new' occurrences in BODY.
+Reccurse if BODY is a list."
+  (if (listp body)
+      (progn
+        (cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate body)
+        nil)
+    (eq body old)))
```

Comment:
`old` is dynamically scoped. It's better to make it a function argument.

Reply:
I thought about it but I could not do that, the predicate for `cl-substitute` is a function with one argument.
What could be done to improve upon this ? Sort of a curried function ? Or maybe declare variables explicitly ? Another macro ?

Reply:
Solution one

well. let's make it to take pair, (OLD . BODY).
it's a function taking ONE argument.
`(cl-nsubstitute-if new 'spacemacs//ruby-substitute-predicate (cons old body)`

---

Make it take two arguments, the first is `old`
Solution two `(cl-nsubstitute-if new (apply-partially old) body)`.